test/Test/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,525 @@
+
+    arbitrary = SerialisedKey' <$>
+                arbitrary `suchThat` (Primitive.length >>> (< 0x10000))
+
+    -- This shrinker relies on lists not becoming longer by shrinking.
+    shrink = shrinkMap SerialisedKey' (\ (SerialisedKey' bytes) -> bytes)
+
+instance Arbitrary IndexOrdinary where
+
+    arbitrary = IndexOrdinary <$> fromList <$> arbitrary
+
+    shrink = shrinkMap (IndexOrdinary . fromList) (toList . fromIndexOrdinary)
+
+newtype SearchableIndex = SearchableIndex IndexOrdinary deriving stock Show
+
+instance Arbitrary SearchableIndex where
+
+    arbitrary = do
+        availableLastKeys <- (getSorted <$> arbitrary) `suchThat` (not . null)
+        lastKeys <- concat <$>
+                    mapM ((<$> genKeyCount) . flip replicate) availableLastKeys
+        return $ SearchableIndex (indexFromLastKeys lastKeys)
+        where
+
+        genKeyCount :: Gen Int
+        genKeyCount = frequency $
+                      [(4, pure 1), (1, getSmall <$> getPositive <$> arbitrary)]
+
+    shrink (SearchableIndex (IndexOrdinary lastKeysVec))
+        = [
+              SearchableIndex (indexFromLastKeys lastKeys') |
+                  lastKeys' <- shrink (toList lastKeysVec),
+                  not (null lastKeys'),
+                  and (zipWith (<=) lastKeys' (List.tail lastKeys'))
+          ]
+
+-- A byte string that is too short to be a serialised index.
+newtype TooShortByteString = TooShortByteString ShortByteString
+    deriving stock Show
+
+fromTooShortByteString :: TooShortByteString -> ShortByteString
+fromTooShortByteString (TooShortByteString shortByteString) = shortByteString
+
+instance Arbitrary TooShortByteString where
+
+    arbitrary = do
+        len <- chooseInt (0, 11)
+        TooShortByteString <$> ShortByteString.pack <$> vector len
+
+    shrink = fromTooShortByteString                      >>>
+             shrink                                      >>>
+             filter (ShortByteString.length >>> (<= 11)) >>>
+             map TooShortByteString
+
+deriving stock instance Show Append
+
+instance Arbitrary Append where
+
+    arbitrary = oneof $
+                [
+                    AppendSinglePage <$> arbitrary <*> arbitrary,
+                    AppendMultiPage <$> arbitrary <*> (getSmall <$> arbitrary)
+                ]
+
+    shrink (AppendSinglePage firstKey lastKey)
+        = [AppendSinglePage firstKey' lastKey | firstKey' <- shrink firstKey] ++
+          [AppendSinglePage firstKey lastKey' | lastKey' <- shrink lastKey]
+    shrink (AppendMultiPage key overflowPageCount)
+        = [AppendSinglePage key key]
+          ++
+          [
+              AppendMultiPage key' overflowPageCount |
+                  key' <- shrink key
+          ]
+          ++
+          [
+              AppendMultiPage key overflowPageCount' |
+                  overflowPageCount' <- shrinkMap getSmall Small $
+                                        overflowPageCount
+          ]
```

Comment:
Another way of doing this (which also the compact index tests use) would be to use the existing generators for `LogicalPageSummaries`, which has a function `toAppends :: _ -> [Append]`.

Reply:
Hmm, not sure about this. It seems to me that using the `LogicalPageSummaries` infrastructure for my tests would be overkill, in particular as that infrastructure deals with first and last keys of pages, while ordinary indexes only maintain last keys. Leaving the above code as it is seems the better alternative to me, at least for the time being.

Why does the compact index actually store both first and last keys? To me, this rather seems like maintaining redundant information, with the potential of making various things, tests for example, more complicated than needed.

Reply:
> Why does the compact index actually store both first and last keys?

It doesn't.

`LogicalPageSummaries` does, because it is a representation of all information available from adding pages to an index. It's isomorphic to `[Append]`, so it seemed redundant to have separate `Arbitrary` instances. The compact index tests also use `LogicalPageSummaries`.

Reply:
I’ve made corresponding changes in 7619f450a8f541fa176edf83974e2259e3c81b1e.